README.md
# NAME

Finance::VolSurface -  represents a volatility surface

# SYNOPSIS

    use feature qw(say);
    use Finance::VolSurface;

    my $volsurface = Finance::VolSurface::Delta->new(
        surface       => { ... },  # see L</Delta surface> for format
        recorded_date => $date,    # this is a L<Date::Utility> instance
        underlying    => Finance::Underlying->by_symbol('frxEURUSD'),
        r_rates       => Finance::YieldCurve->new(asset => 'EUR', data => { ... }),
        q_rates       => Finance::YieldCurve->new(asset => 'USD', data => { ... }),
    );

    # Interpolate points on the surface to get a single number for volatility
    my $vol = $volsurface->get_volatility(
        delta => 50,
        from  => $now,  # This is a L<Date::Utility> instance
        to    => $now->plus('3d'),
    );

    # TODO - Spread from max or atm
    my $spread = $volsurface->get_spread(
        sought_point => 'atm', # may rename to delta
        days         => 7,     # may rename to tenor
    );

    # Validation for the supplied volsurface
    die 'incorrect volsurface provided: ' . $volsurface->validation_error unless $volsurface->is_valid;

# DESCRIPTION

## Delta surface

Raw surface data for a delta surface:

    {
      'ON' => {
          smile => {
              50 => 0.4,
              25 => 0.2,
              75 => 0.7,
          },
          spread => {
              50 => 0.1,
              25 => 0.1,
              75 => 0.1,
          },
      },
      '1W' => {
          smile => {
              50 => 0.4,
              25 => 0.2,
              75 => 0.7,
          },
          spread => {
              50 => 0.1,
              25 => 0.1,
              75 => 0.1,
          },
      },
      '2W' => { ... }
    }

Expected tenors could include:

\* ON for overnight
\* 1W for 1 week
\* 6M for 6 month
\* 1Y for 1 year

Internally, the key for the surface is always a number of days (the tenor),
and for overnight this would typically be 1 to 3 (for weekends).

## Moneyness

The keys in the smile hashref are moneyness points as percentages (100 = 100%),
typically ranging from 80%-120%.

Spread has a single atm value.

    {
      1 => {
          smile => {
              80 => 0.2,
              82 => 0.2,
              84 => 0.2,
              88 => 0.2,
              92 => 0.2,
              96 => 0.2,
              100 => 0.4,
              102 => 0.4,
              104 => 0.4,
              108 => 0.4,
              114 => 0.4,
              120 => 0.7,
          },
          spread => {
              100 => 0.1,
          },
      },
      7 => { ... },
    }

## Flat

This is a single point.

    {
      1 => {
          smile => {
              100 => 0.1,
          },
          spread => {
              100 => 0,
          },
      },
      7 => { ... },
    }

## Construction

Note that a volsurface instance must always be created from the appropriate subclass,
i.e. one of:

- [Finance::VolSurface::Delta](https://metacpan.org/pod/Finance::VolSurface::Delta)
- [Finance::VolSurface::Moneyness](https://metacpan.org/pod/Finance::VolSurface::Moneyness)
- [Finance::VolSurface::Flat](https://metacpan.org/pod/Finance::VolSurface::Flat)

# ATTRIBUTES

## effective\_date

Surfaces roll over at 5pm NY time, so the vols of any surfaces recorded after 5pm NY but
before GMT midnight are effectively for the next GMT day. This attribute holds this
effective date.

## recorded\_date

The date (and time) that the surface was recorded, as a [Date::Utility](https://metacpan.org/pod/Date::Utility). This should
be provided on construction.

## q\_rates

A yield curve for Q rates. For stocks, these would represent dividends.

## r\_rates

A yield curve for R rates. This would typically be interest rates.

## smile\_points

The points across a smile.

It can be delta points, moneyness points or any other points that we might have in the future.

Returns an arrayref of numerical point values that comprise the smile.

## spread\_points

This will give an array-reference containing volatility spreads for first tenor which has a volatility spread (or ATM if none).

## surface

Volatility surface in a hash reference.

## surface\_data

The original surface data.

## symbol

The symbol of the underlying that this surface is for (e.g. frxUSDJPY)

## term\_by\_day

Get all the terms in a surface in ascending order.

## type

Type of the surface, delta, moneyness or flat.

## underlying

The [Finance::Underlying](https://metacpan.org/pod/Finance::Underlying) for this volsurface (mandatory).

## get\_rr\_bf\_for\_smile

Return the rr and bf values for a given smile
For more info see: https://en.wikipedia.org/wiki/Risk\_reversal and https://en.wikipedia.org/wiki/Butterfly\_(options)

## get\_surface\_smile

Returns the smile on the surface.
Returns an empty hash reference if not present.

## is\_valid

Does this volatility surface pass our validation.

## get\_volatility

Calculates volatility from the surface based input parameters.

Expects 3 mandatory arguments as input.

1) from - Date::Utility object
2) to - Date::Utility object
3) delta | strike | moneyness.

For a moneyness surface, the `spot` value is also required.

Will return a single volatility value, or throw an exception if the volsurface or parameters
are invalid.

Examples:

    my $from = Date::Utility->new('2016-06-01 10:00:00');
    my $to   = Date::Utility->new('2016-06-01 15:00:00');
    my $vol  = $s->get_volatility({delta => 25, from => $from, to => $to});
    my $vol  = $s->get_volatility({strike => $bet->barrier, from => $from, to => $to});
    my $vol  = $s->get_volatility({moneyness => 95, spot => 104.23, from => $from, to => $to});

## atm\_spread\_point

(to be defined)

## variance\_table

A variance surface. Converted from raw volatility input surface.
Only available on delta volsurfaces.

## get\_smile

Calculate the requested smile from volatility surface.

Usage:

    my $smile = $vol_surface->get_smile($days);

## get\_variances

Calculate the variance for a given date based on volatility surface data.

Only applicable to delta volsurfaces.

## get\_weight

Get the weight between two given dates.

## get\_market\_rr\_bf

Returns the rr and bf values for a given day

## get\_smile\_expiries

An array reference of that contains expiry dates for smiles on the volatility surface.

## min\_vol\_spread

minimum volatility spread that we can accept for this volatility surface.

## interpolate

Quadratic interpolation to interpolate across smile

    $surface->interpolate({smile => $smile, sought_point => $sought_point});

---

# NAME

Finance::VolSurface::Delta

# DESCRIPTION

See [Finance::VolSurface](https://metacpan.org/pod/Finance::VolSurface).

---

# NAME

Finance::VolSurface::ExpiryConventions - utilities for dealing with expiry for converting tenor to/from a number of days

# SYNOPSIS

    use Finance::VolSurface::ExpiryConventions;

# DESCRIPTION

---

# NAME

Finance::VolSurface::Moneyness

# DESCRIPTION

See [Finance::VolSurface](https://metacpan.org/pod/Finance::VolSurface).

---

## finance\_volsurface\_type

Volatility surface types.

---

# NAME

Finance::VolSurface::Utils

# DESCRIPTION

Some general vol-related utility functions.

# SYNOPSIS

## NY1700\_rollover\_date\_on

Returns (as a Date::Utility) the NY1700 rollover date for a given Date::Utility.

## effective\_date\_for

Get the "effective date" for a given Date::Utility (stated in GMT).

This is the we should consider a volsurface effective for, and rolls over
every day at NY1700. If a volsurface is quoted at GMT2300, its effective
date is actually the next day.

This returns a Date::Utility truncated to midnight of the relevant day.

## is\_before\_rollover

Returns 1 if given date-time is before roll-over time.

## get\_ny\_offset\_from\_gmt

Returns offset in hours for the given epoch for NY vs GMT.
Caches output per hour.


Changes
Revision history for Finance-VolSurface

{{$NEXT}}


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.48;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Abstraction for dealing with volatility surfaces",
  "AUTHOR" => "binary.com <cpan\@binary.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.48"
  },
  "DISTNAME" => "Finance-VolSurface",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Finance::VolSurface",
  "PREREQ_PM" => {
    "Date::Utility" => 0,
    "Format::Util::Numbers" => 0,
    "Math::Util::CalculatedValue::Validatable" => 0,
    "Moose" => 0,
    "Time::Duration::Concise" => 0,
    "Time::HiRes" => 0,
    "YAML" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Date::Utility" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Format::Util::Numbers" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Math::Util::CalculatedValue::Validatable" => 0,
  "Moose" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.94",
  "Time::Duration::Concise" => 0,
  "Time::HiRes" => 0,
  "YAML" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Finance/VolSurface/Utils.pm
package Finance::VolSurface::Utils;

=head1 NAME

Finance::VolSurface::Utils

=head1 DESCRIPTION

Some general vol-related utility functions.

=head1 SYNOPSIS

=cut

no indirect;
use Moo;

use DateTime::TimeZone;
use Date::Utility;
use Number::Closest::XS qw(find_closest_numbers_around);
use List::MoreUtils qw(notall);
use List::Util qw(min max);

## VERSION

=head2 NY1700_rollover_date_on

Returns (as a Date::Utility) the NY1700 rollover date for a given Date::Utility.

=cut

sub NY1700_rollover_date_on {
    my ($self, $date) = @_;

    return $date->truncate_to_day->plus_time_interval((17 - $date->timezone_offset('America/New_York')->hours) * 3600);
}

=head2 effective_date_for

Get the "effective date" for a given Date::Utility (stated in GMT).

This is the we should consider a volsurface effective for, and rolls over
every day at NY1700. If a volsurface is quoted at GMT2300, its effective
date is actually the next day.

This returns a Date::Utility truncated to midnight of the relevant day.

=cut

sub effective_date_for {
    my ($self, $date) = @_;

    return $date->plus_time_interval((7 + $date->timezone_offset('America/New_York')->hours) * 3600)->truncate_to_day;
}

=head2 is_before_rollover

Returns 1 if given date-time is before roll-over time.

=cut

sub is_before_rollover {
    my ($self, $date) = @_;

    return ($date->is_after($self->NY1700_rollover_date_on($date))) ? 0 : 1;
}

sub _get_points_to_interpolate {
    my ($seek, $available_points) = @_;
    die('Need 2 or more term structures to interpolate.')
        if scalar @$available_points <= 1;

    return @{find_closest_numbers_around($seek, $available_points, 2)};
}

sub _is_between {
    my ($seek, $points) = @_;

    my @points = @$points;

    die('some of the points are not defined')
        if (notall { defined $_ } @points);
    die('less than two available points')
        if (scalar @points < 2);

    return if $seek > max(@points) or $seek < min(@points);
    return 1;
}

=head2 get_ny_offset_from_gmt

Returns offset in hours for the given epoch for NY vs GMT.
Caches output per hour.

=cut

my $ny_offset_from_gmt_by_hour = {};

sub get_ny_offset_from_gmt {
    my $epoch = shift;
    $epoch = int($epoch / 3600);
    $ny_offset_from_gmt_by_hour->{$epoch} = Date::Utility->new($epoch * 3600)->timezone_offset('America/New_York')->hours
        unless $ny_offset_from_gmt_by_hour->{$epoch};
    return $ny_offset_from_gmt_by_hour->{$epoch};
}

1;


lib/Finance/VolSurface.pm
package Finance::VolSurface;
# ABSTRACT: Abstraction for dealing with volatility surfaces

use strict;
use warnings;

our $VERSION = '0.001';

=head1 NAME

Finance::VolSurface -  represents a volatility surface

=head1 SYNOPSIS

    use feature qw(say);
    use Finance::VolSurface;

    my $volsurface = Finance::VolSurface::Delta->new(
        surface       => { ... },  # see L</Delta surface> for format
        recorded_date => $date,    # this is a L<Date::Utility> instance
        underlying    => Finance::Underlying->by_symbol('frxEURUSD'),
        r_rates       => Finance::YieldCurve->new(asset => 'EUR', data => { ... }),
        q_rates       => Finance::YieldCurve->new(asset => 'USD', data => { ... }),
    );

    # Interpolate points on the surface to get a single number for volatility
    my $vol = $volsurface->get_volatility(
        delta => 50,
        from  => $now,  # This is a L<Date::Utility> instance
        to    => $now->plus('3d'),
    );

    # TODO - Spread from max or atm
    my $spread = $volsurface->get_spread(
        sought_point => 'atm', # may rename to delta
        days         => 7,     # may rename to tenor
    );

    # Validation for the supplied volsurface
    die 'incorrect volsurface provided: ' . $volsurface->validation_error unless $volsurface->is_valid;

=head1 DESCRIPTION

=head2 Delta surface

Raw surface data for a delta surface:

 {
   'ON' => {
       smile => {
           50 => 0.4,
           25 => 0.2,
           75 => 0.7,
       },
       spread => {
           50 => 0.1,
           25 => 0.1,
           75 => 0.1,
       },
   },
   '1W' => {
       smile => {
           50 => 0.4,
           25 => 0.2,
           75 => 0.7,
       },
       spread => {
           50 => 0.1,
           25 => 0.1,
           75 => 0.1,
       },
   },
   '2W' => { ... }
 }

Expected tenors could include:

* ON for overnight
* 1W for 1 week
* 6M for 6 month
* 1Y for 1 year

Internally, the key for the surface is always a number of days (the tenor),
and for overnight this would typically be 1 to 3 (for weekends).

=head2 Moneyness

The keys in the smile hashref are moneyness points as percentages (100 = 100%),
typically ranging from 80%-120%.

Spread has a single atm value.

 {
   1 => {
       smile => {
           80 => 0.2,
           82 => 0.2,
           84 => 0.2,
           88 => 0.2,
           92 => 0.2,
           96 => 0.2,
           100 => 0.4,
           102 => 0.4,
           104 => 0.4,
           108 => 0.4,
           114 => 0.4,
           120 => 0.7,
       },
       spread => {
           100 => 0.1,
       },
   },
   7 => { ... },
 }

=head2 Flat

This is a single point.

 {
   1 => {
       smile => {
           100 => 0.1,
       },
       spread => {
           100 => 0,
       },
   },
   7 => { ... },
 }

=head2 Construction

Note that a volsurface instance must always be created from the appropriate subclass,
i.e. one of:

=over 4

=item * L<Finance::VolSurface::Delta>

=item * L<Finance::VolSurface::Moneyness>

=item * L<Finance::VolSurface::Flat>

=back

=cut

no indirect;

use Moose;

use Date::Utility;
use Try::Tiny;

use List::Util qw(first);
use List::MoreUtils qw(notall any);
use List::Util qw( min max first );

use Finance::Underlying;
use Finance::YieldCurve;

use Finance::VolSurface::Utils;
use Finance::VolSurface::Types qw(Finance_VolSurface_Type);
use Finance::VolSurface::ExpiryConventions;

our $extra_vol_diff_by_delta = 0.1;

=head1 ATTRIBUTES

=head2 effective_date

Surfaces roll over at 5pm NY time, so the vols of any surfaces recorded after 5pm NY but
before GMT midnight are effectively for the next GMT day. This attribute holds this
effective date.

=cut

has effective_date => (
    is         => 'ro',
    init_arg   => undef,
    lazy_build => 1,
);

sub _build_effective_date {
    my $self = shift;

    return Finance::VolSurface::Utils->new->effective_date_for($self->recorded_date);
}

=head2 recorded_date

The date (and time) that the surface was recorded, as a L<Date::Utility>. This should
be provided on construction.

=cut

has recorded_date => (
    is         => 'ro',
    lazy_build => 1,
);

=head2 q_rates

A yield curve for Q rates. For stocks, these would represent dividends.

=cut

has q_rates => (
    is       => 'ro',
    isa      => 'Finance::YieldCurve',
    required => 1,
);

=head2 r_rates

A yield curve for R rates. This would typically be interest rates.

=cut

has r_rates => (
    is       => 'ro',
    isa      => 'Finance::YieldCurve',
    required => 1,
);

=head2 smile_points

The points across a smile.

It can be delta points, moneyness points or any other points that we might have in the future.

Returns an arrayref of numerical point values that comprise the smile.

=cut

has smile_points => (
    is         => 'ro',
    isa        => 'ArrayRef',
    lazy_build => 1,
);

sub _build_smile_points {
    my $self = shift;

    # Default to the point found in the first day we find
    # in $self->surface that has a smile. As long as each smile
    # has the same points, this works. If each smile has different
    # points, the Validator is going to give you trouble!
    my $surface = $self->surface;
    my $suitable_day = first { exists $surface->{$_}->{smile} } @{$self->term_by_day};

    return [sort { $a <=> $b } keys %{$surface->{$suitable_day}->{smile}}] if $suitable_day;
    return [];
}

=head2 spread_points

This will give an array-reference containing volatility spreads for first tenor which has a volatility spread (or ATM if none).

=cut

has spread_points => (
    is         => 'ro',
    isa        => 'ArrayRef',
    lazy_build => 1,
);

sub _build_spread_points {
    my $self = shift;

    # Default to the point found in the first day we find
    # in $self->surface that has a volspread. As long as each volspread
    # has the same points, this works. If each smile has different
    # points, the Validator is going to give you trouble!
    my $surface = $self->surface;
    my $suitable_day = first { exists $surface->{$_}->{vol_spread} } keys %{$surface};

    return [sort { $a <=> $b } keys %{$surface->{$suitable_day}{vol_spread}}] if $suitable_day;
    return [];
}

=head2 surface

Volatility surface in a hash reference.

=cut

has surface => (
    is         => 'ro',
    lazy_build => 1,
);

=head2 surface_data

The original surface data.

=cut

has surface_data => (
    is         => 'ro',
    lazy_build => 1,
);

=head2 symbol

The symbol of the underlying that this surface is for (e.g. frxUSDJPY)

=cut

sub symbol { shift->underlying->symbol }

=head2 term_by_day

Get all the terms in a surface in ascending order.

=cut

has term_by_day => (
    is         => 'ro',
    isa        => 'ArrayRef',
    init_arg   => undef,
    lazy_build => 1,
);

sub _build_term_by_day {
    my $self = shift;

    return [sort { $a <=> $b } keys %{$self->surface}];
}

=head2 type

Type of the surface, delta, moneyness or flat.

=cut

has type => (
    is       => 'ro',
    isa      => Finance_VolSurface_Type,
    required => 1,
    init_arg => undef,
    default  => undef,
);

=head2 underlying

The L<Finance::Underlying> for this volsurface (mandatory).

=cut

has underlying => (
    is  => 'ro',
    isa => 'Finance::Underlying',
);

=head2 get_rr_bf_for_smile

Return the rr and bf values for a given smile
For more info see: https://en.wikipedia.org/wiki/Risk_reversal and https://en.wikipedia.org/wiki/Butterfly_(options)

=cut

sub get_rr_bf_for_smile {
    my ($self, $market_smile) = @_;

    my $result = {
        ATM   => $market_smile->{50},
        RR_25 => $market_smile->{25} - $market_smile->{75},
        BF_25 => ($market_smile->{25} + $market_smile->{75}) / 2 - $market_smile->{50},
    };
    if (exists $market_smile->{10}) {
        $result->{RR_10} = $market_smile->{10} - $market_smile->{90};
        $result->{BF_10} = ($market_smile->{10} + $market_smile->{90}) / 2 - $market_smile->{50};
    }

    return $result;
}

=head2 get_surface_smile

Returns the smile on the surface.
Returns an empty hash reference if not present.

=cut

sub get_surface_smile {
    my ($self, $days) = @_;

    return $self->surface->{$days}->{smile} // {};
}

=head2 is_valid

Does this volatility surface pass our validation.

=cut

sub is_valid {
    my $self = shift;

    try {
        $self->$_ for ($self->_validation_methods);
    }
    catch {
        $self->validation_error($_);
    };

    return !$self->validation_error;
}

sub _validate_age {
    my $self = shift;

    if (time - $self->recorded_date->epoch > 4 * 3600) {
        die('Volatility surface from provider for ' . $self->symbol . ' is more than 4 hours old.');
    }

    return;
}

sub _validate_structure {
    my $self = shift;

    my $surface_hashref = $self->surface;
    my $system_symbol   = $self->symbol;

    # Somehow I do not know why there is a limit of term on delta surface, but
    # for moneyness we might need at least up to 2 years to get the spread.
    my $extra_allowed = $Quant::Framework::Underlying::extra_vol_diff_by_delta || 0;
    my $max_vol_change_by_delta = 0.4 + $extra_allowed;

    my @days = sort { $a <=> $b } keys %{$surface_hashref};

    if (@days < 2) {
        die('Must be at least two maturities on vol surface for ' . $self->symbol . '.');
    }

    if ($days[-1] > $self->_max_allowed_term) {
        die("Day[$days[-1]] in volsurface for underlying[$system_symbol] greater than allowed[" . $self->_max_allowed_term . "].");
    }

    if ($self->underlying->market->name eq 'forex' and $days[0] > 7) {
        die("ON term is missing in volsurface for underlying $system_symbol, the minimum term is $days[0].");
    }

    foreach my $day (@days) {
        if ($day !~ /^\d+$/) {
            die("Invalid day[$day] in volsurface for underlying[$system_symbol]. Not a positive integer.");
        }
    }

    foreach my $day (grep { exists $surface_hashref->{$_}->{smile} } @days) {
        my $smile = $surface_hashref->{$day}->{smile};
        my @volatility_level = sort { $a <=> $b } keys %$smile;

        for (my $i = 0; $i < $#volatility_level; $i++) {
            my $level = $volatility_level[$i];

            if ($level !~ /^\d+\.?\d+$/) {
                die("Invalid vol_point[$level] for underlying[$system_symbol].");
            }

            my $next_level = $volatility_level[$i + 1];
            if (abs($level - $next_level) > $self->_max_difference_between_smile_points) {
                die("Difference between point $level and $next_level is too great for days $day.");
            }

            if (not $self->_is_valid_volatility_smile($smile)) {
                die("Invalid smile volatility on $day for $system_symbol");
            }

            if (abs($smile->{$level} - $smile->{$next_level}) > $max_vol_change_by_delta * $smile->{$level}) {
                die(      "Invalid volatility points: too big jump from "
                        . "$level:$smile->{$level} to $next_level:$smile->{$next_level}"
                        . "for maturity[$day], underlying[$system_symbol]");
            }
        }
    }

    return;
}

sub _is_valid_volatility_smile {
    my ($self, $smile) = @_;

    foreach my $vol (values %$smile) {
        # sanity check on volatility. Cannot be more than 500% and must be a number.
        return if ($vol !~ /^\d?\.?\d*$/ or $vol > 5);
    }

    return 1;
}

sub _validate_identical_surface {
    my $self = shift;

    my $existing = $self->new({
        underlying       => $self->underlying,
        chronicle_reader => $self->chronicle_reader,
        chronicle_writer => $self->chronicle_writer,
    });

    my $existing_surface_data = $existing->surface;
    my $new_surface_data      = $self->surface;

    my @existing_terms = sort { $a <=> $b } grep { exists $existing_surface_data->{$_}{smile} } keys %$existing_surface_data;
    my @new_terms      = sort { $a <=> $b } grep { exists $new_surface_data->{$_}{smile} } keys %$new_surface_data;

    return if @existing_terms != @new_terms;
    return if any { $existing_terms[$_] != $new_terms[$_] } (0 .. $#existing_terms);

    foreach my $term (@existing_terms) {
        my $existing_smile = $existing_surface_data->{$term}->{smile};
        my $new_smile      = $new_surface_data->{$term}->{smile};
        return if (scalar(keys %$existing_smile) != scalar(keys %$new_smile));
        return if (any { $existing_smile->{$_} != $new_smile->{$_} } keys %$existing_smile);
    }

    my $existing_surface_epoch = Date::Utility->new($self->document->{date})->epoch;

    if (time - $existing_surface_epoch > 15000 and not $self->underlying->quanto_only) {
        die('Surface data has not changed since last update [' . $existing_surface_epoch . '] for ' . $self->symbol . '.');
    }

    return;
}

sub _validate_volatility_jumps {
    my $self = shift;

    my $existing = $self->new({
        underlying       => $self->underlying,
        chronicle_reader => $self->chronicle_reader,
        chronicle_writer => $self->chronicle_writer,
    });

    my @terms           = @{$self->original_term_for_smile};
    my @new_expiry      = @{$self->get_smile_expiries};
    my @existing_expiry = @{$existing->get_smile_expiries};

    my @points = @{$self->smile_points};
    my $type   = $self->type;

    for (my $i = 1; $i <= $#new_expiry; $i++) {
        for (my $j = 0; $j <= $#points; $j++) {
            my $sought_point = $points[$j];
            my $new_vol      = $self->get_volatility({
                $type => $sought_point,
                from  => $self->recorded_date,
                to    => $new_expiry[$i],
            });
            my $existing_vol = $existing->get_volatility({
                $type => $sought_point,
                from  => $existing->recorded_date,
                to    => $existing_expiry[$i],
            });
            my $diff = abs($new_vol - $existing_vol);
            if ($diff > 0.03 and $diff > $existing_vol) {
                die('Big difference found on term[' . $terms[$i - 1] . '] for point [' . $sought_point . '] with absolute diff [' . $diff . '].');
            }
        }
    }

    return;
}

has validation_error => (
    is      => 'rw',
    default => '',
);

=head2 get_volatility

Calculates volatility from the surface based input parameters.

Expects 3 mandatory arguments as input.

1) from - Date::Utility object
2) to - Date::Utility object
3) delta | strike | moneyness.

For a moneyness surface, the C<spot> value is also required.

Will return a single volatility value, or throw an exception if the volsurface or parameters
are invalid.

Examples:

  my $from = Date::Utility->new('2016-06-01 10:00:00');
  my $to   = Date::Utility->new('2016-06-01 15:00:00');
  my $vol  = $s->get_volatility({delta => 25, from => $from, to => $to});
  my $vol  = $s->get_volatility({strike => $bet->barrier, from => $from, to => $to});
  my $vol  = $s->get_volatility({moneyness => 95, spot => 104.23, from => $from, to => $to});

=cut

sub get_volatility { ... }

=head2 atm_spread_point

(to be defined)

=cut

has atm_spread_point => (
    is      => 'ro',
    isa     => 'Num',
);

=head2 variance_table

A variance surface. Converted from raw volatility input surface.
Only available on delta volsurfaces.

=cut

has variance_table => (
    is         => 'ro',
    lazy_build => 1,
);

=head2 get_smile

Calculate the requested smile from volatility surface.

Usage:

    my $smile = $vol_surface->get_smile($days);

=cut

sub get_smile { ... }

=head2 get_variances

Calculate the variance for a given date based on volatility surface data.

Only applicable to delta volsurfaces.

=cut

sub get_variances { ... }

=head2 get_weight

Get the weight between two given dates.

=cut

sub get_weight { ... }

=head2 get_market_rr_bf

Returns the rr and bf values for a given day

=cut

=head2 get_smile_expiries

An array reference of that contains expiry dates for smiles on the volatility surface.

=cut

=head2 min_vol_spread

minimum volatility spread that we can accept for this volatility surface.

=cut

has min_vol_spread => (
    is      => 'ro',
    isa     => 'Num',
    default => 3.1 / 100,
);

=head2 interpolate

Quadratic interpolation to interpolate across smile

    $surface->interpolate({smile => $smile, sought_point => $sought_point});

=cut

1;

